module-apps/apps-common/widgets/spinners/Container.hpp

```diff
@@ -0,0 +1,226 @@
+// For licensing, see https://github.com/mudita/MuditaOS/LICENSE.md
+
+#pragma once
+
+#include <vector>
+#include <string>
+#include <type_traits>
+
+/// Generic elements container.
+/// By default, the container will choose the most optimal way of elements storage.
+/// Fundamental types(ints, float, doubles, chars) are stored as simple variables. Compound types are stored in
+/// std::vector.
+/// Sometimes it is necessary to force the latter way of storing values even for fundamental types. For
+/// instance, the user might want to store a fixed list of integers to iterate. In that case, set the force parameter to
+/// true.
+template <class ElementType, bool force = false, typename = void> class Container
+{
+  public:
+    enum class Boundaries
+    {
+        Fixed,     /// Stop scrolling upon reaching last/first element
+        Continuous /// Jump to beginning/end upon reaching last/first element
+    };
+    using range      = std::vector<ElementType>;
+    using value_type = ElementType;
+
+    Container(range &&elements, Boundaries boundaries) : elements{std::move(elements)}, boundaries{boundaries}
+    {}
+
+    Container(const range &elements, Boundaries boundaries) : elements{elements}, boundaries{boundaries}
+    {}
+
+    ElementType get() const
+    {
+        try {
+            return elements.at(pos);
+        }
+        catch (...) {
+            return {};
+        }
+    }
+
+    void set(ElementType val)
+    {
+        for (auto i = 0U; i < elements.size(); i++) {
+            if (elements[i] == val) {
+                pos = i;
+                break;
+            }
+        }
+    }
+
+    bool next()
+    {
+        bool ret{true};
+        if (pos >= range_top()) {
+            if (boundaries == Boundaries::Continuous) {
+                pos = 0;
+            }
+            else {
+                pos = range_top();
+                ret = false;
+            }
+        }
+        else {
+            pos++;
+        }
+        return ret;
+    }
+
+    bool previous()
+    {
+        bool ret{true};
+        if (pos <= 0) {
+            if (boundaries == Boundaries::Continuous) {
+                pos = range_top();
+            }
+            else {
+                pos = 0;
+                ret = false;
+            }
+        }
+        else {
+            pos--;
+        }
+        return ret;
+    }
+
+    void set_range(range newRange)
+    {
+        if (elements != newRange) {
+            elements = newRange;
+            pos      = 0;
+        }
+    }
+
+    [[nodiscard]] size_t size() const
+    {
+        return elements.size();
+    }
+
+    [[nodiscard]] bool is_min() const
+    {
+        return pos == 0;
+    }
+
+    [[nodiscard]] bool is_max() const
+    {
+        return pos == range_top();
+    }
+
+  private:
+    [[nodiscard]] std::uint32_t range_top() const
+    {
+        return elements.size() - 1;
+    }
+
+    range elements;
+    std::uint32_t pos{};
```

Comment:
We work on stl containers here - why not iterator? 
Also... is this logic really required - can't we have derived "rotary" or not "rotary" which would just change `next` logic? for rotary you could use `std::rotate` most probably :)

Reply:
Of course we can and it would be better than the current solution as we don't need to change this behavior at runtime. Unfortunately, there is a lot of code that is using such an approach already. As you might already know this would trigger more and more changes. IMHO this is a good proposal. Why not create a jira task for it?